### PR TITLE
[Internal Manager] Fix not triggering transactions by migrating to wagmi viem

### PR DESCRIPTION
**Pull request by luizakp**
Followed: 
- [Rainbowkit migration guide](https://www.rainbowkit.com/docs/migration-guide)
- [Wagmi migration guide](https://wagmi.sh/core/migration-guide)

Changed:
- removed `usePrepareVaultManageUserBalance` and `useVaultManageUserBalance` and used async functions instead. This removes the state update and calls the contract function directly on submit

https://www.loom.com/share/4ca8efc98fb54ba29056e305571557dc?sid=1b043dad-975b-403b-a5dc-92a5ff7ed1f7

### CODE REVIEW

File: packages/next-app/next.config.js [hunk 1]
```diff
@@ -12,6 +12,10 @@ const moduleExports = {
   },
   reactStrictMode: true,
   swcMinify: true,
+  webpack: config => {
+    config.resolve.fallback = { fs: false, net: false, tls: false };
+    return config;
+  },
```
Comment (ribeirojose) [inline packages/next-app/next.config.js:18]: can you elaborate on why we need this? 🤔

File: packages/next-app/src/polyfills.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { Buffer } from "buffer";
```
Comment (ribeirojose) [inline packages/next-app/src/polyfills.ts:1]: if this is also related to wagmi can you leave a comment in the top of the file explaining this?

File: packages/next-app/src/hooks/useTransaction.ts [hunk 3]
```diff
@@ -181,19 +200,19 @@ export function useMetadataTransaction({
       setNotification(NOTIFICATION_MAP[TransactionStatus.PINNING]);
 
       // Call function to approve transaction and pin metadata to IPFS here
-      // Once the transaction is approved and the metadata is pinned, update the transaction status to CONFIRMING
+      // Once the transaction is approved and the metadata is pinned, update the transaction status to APPROVED
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/useTransaction.ts:203]: 👍

File: packages/next-app/src/hooks/useTransaction.ts [hunk 4]
```diff
@@ -294,118 +293,126 @@ export function useInternalBalancesTransaction({
...
-
-  useEffect(() => {
-    if (submitData.length === 0) return;
+  async function handleTransaction({ data }: { data: SubmitData[] }) {
     setTransactionUrl(undefined);
-    setNotification(
-      NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.AUTHORIZING]
-    );
     if (
       operationKind === UserBalanceOpKind.DEPOSIT_INTERNAL &&
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/useTransaction.ts:347]: 😬

File: packages/next-app/src/wagmi/client.ts [hunk 5]
```diff
@@ -9,9 +11,8 @@ import {
   polygonZkEvm as polygonZkEvmChain,
   sepolia,
 } from "wagmi/chains";
-import { publicProvider } from "wagmi/providers/public";
 
-import { configureChains, createClient } from "#/wagmi";
+const projectId = "4f98524b2b9b5a80d14a519a8dcbecc2";
```
Comment (ribeirojose) [inline packages/next-app/src/wagmi/client.ts:15]: Where's exactly this coming from? Again, something that must absolutely be explaining in the code.

File: packages/next-app/src/wagmi/setPoolMetadata.ts [hunk 6]
```diff
@@ -1,25 +0,0 @@
-import { Address } from "@bleu-balancer-tools/shared";
```
Comment (ribeirojose) [inline packages/next-app/src/wagmi/setPoolMetadata.ts:1]: I gather this was simply not being used anymore?

File: packages/next-app/src/hooks/useTransaction.ts [hunk 7]
```diff
@@ -294,118 +293,126 @@ export function useInternalBalancesTransaction({
...
+            sender: userAddress,
+            recipient: transactionItem.receiverAddress as Address,
+            asset: transactionItem.tokenAddress,
+            amount: parseUnits(
+              transactionItem.tokenAmount,
+              transactionItem.tokenDecimals
+            ),
+          };
+        });
+        const config = await prepareWriteContract({
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/useTransaction.ts:372]: I know this was already here but want to reiterate my suggestion to keep domain-related stuff somewhere else. This hook is so complex already -- adding domain-specific code just makes it even harder to reason about it.

**Review by ribeirojose — DISMISSED** [review]
Thanks for taking the lead here, left a few comments. Have you run any tests in the metadata app? I think that'd be the only other requirement here.

File: packages/next-app/next.config.js (hunk 1 above)
Comment (luizakp) [reply packages/next-app/next.config.js:18]: this is actually related to rainbowkit migration [3. Ensure bundler and polyfill compatibility](https://www.rainbowkit.com/docs/migration-guide)

Will write a comment about this there

File: packages/next-app/src/polyfills.ts (hunk 2 above)
Comment (luizakp) [reply packages/next-app/src/polyfills.ts:1]: yup!

File: packages/next-app/src/wagmi/setPoolMetadata.ts (hunk 6 above)
Comment (luizakp) [reply packages/next-app/src/wagmi/setPoolMetadata.ts:1]: yes, we are using the wagmi generated hook for a while now. wrote it on the commit message https://github.com/bleu-studio/balancer-tools/pull/157/commits/16ffa083fb5f65ad84811f67fb213b19d1eb7e51

File: packages/next-app/src/hooks/useTransaction.ts (hunk 7 above)
Comment (luizakp) [reply packages/next-app/src/hooks/useTransaction.ts:372]: in this case, the domain-related would be calling the contract?

**Comment by luizakp** [comment]
> Thanks for taking the lead here, left a few comments. Have you run any tests in the metadata app? I think that'd be the only other requirement here.

@ribeirojose did the usual test on metadata. Is there something different than this? https://www.loom.com/share/2d24d43a06914477a7326971368a2276?sid=03fa5a70-cd6a-494f-bd5f-cf17404bf68e

**Comment by luizakp** [comment]
@ribeirojose on https://github.com/bleu-studio/balancer-tools/pull/157/commits/61476426bc684e873d007d3f9968d133b8e8bbc7

I've broken `useInternalBalancesTransaction` into:
- `useCheckAllowance`
- `useTokenApproval`
- `useManageUserBalance`
- `useTransactionStatus`

And for the contracts interactions I've used separate functions (these functions all use @wagmi/core async functions instead wagmi generated hooks because those require dynamic args through state, which creates an unwanted behavior with On Submit forms):
- `readTokenAllowance`
- `writeManageUserBalance`
- `writeTokenApproval`

**Comment by ribeirojose** [comment]
> > Thanks for taking the lead here, left a few comments. Have you run any tests in the metadata app? I think that'd be the only other requirement here.
> 
> 
> 
> @ribeirojose did the usual test on metadata. Is there something different than this? https://www.loom.com/share/2d24d43a06914477a7326971368a2276?sid=03fa5a70-cd6a-494f-bd5f-cf17404bf68e

No, that’s it! Looking sharp 👌

File: packages/next-app/src/hooks/internalmanager/useCheckAllowance.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      return;
+    }
+
+    const allowance = await readTokenAllowance({
+      tokenAddress,
+      userAddress,
+    });
+
+    const amountToApprove = parseUnits(tokenAmount, tokenDecimals);
+    if (allowance >= amountToApprove) {
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/internalmanager/useCheckAllowance.ts:34]: 👌

**Review by ribeirojose — APPROVED** [review]
Looks great!

File: packages/next-app/src/wagmi/client.ts [hunk 9]
```diff
@@ -9,9 +11,14 @@ import {
...
-import { publicProvider } from "wagmi/providers/public";
 
-import { configureChains, createClient } from "#/wagmi";
+/**
+ * Project ID is required by Rainbowkit Migration Guide to Viem
+ * 2. Supply a WalletConnect Cloud projectId
+ * https://www.rainbowkit.com/docs/migration-guide#2-supply-a-walletconnect-cloud-projectid
+ */
+
+const projectId = "4f98524b2b9b5a80d14a519a8dcbecc2";
```
[the remark this answers is not in the document]
Comment (luizakp) [reply packages/next-app/src/wagmi/client.ts:21]: I've created the 1PW vault for Balancer tools but I don't have the role to share with everyone on the team, but the Owners and Administrators do. And sure, I'll comment on that too

[the remark this answers is not in the document]
Comment (ribeirojose) [reply packages/next-app/src/wagmi/client.ts:21]: great, thanks! FWIW just upped everyone to admin

**Review by ribeirojose — APPROVED** [review]

File: packages/next-app/src/wagmi/client.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (luizakp) [reply packages/next-app/src/wagmi/client.ts:22]: moved to `[ext] Balancer` Vault 🙇‍♀️

**Pull request merged**